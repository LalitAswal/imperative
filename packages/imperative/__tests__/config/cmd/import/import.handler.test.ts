/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import ImportHandler from "../../../../src/config/cmd/import/import.handler";
import { URL } from "url";
import { join } from "path";
import * as fs from "fs";

const flushPromises = () => new Promise(setImmediate);

const localSchemaDir: string = join(__dirname, "__resources__");
const localSchema: string = join(localSchemaDir, "zowe.schema.json");
const localDownloadedSchemaPath = join(localSchemaDir, "__data__", "downloaded.schema.json");
const localSchemaUrl: URL = new URL("file://" + localSchema);

const configUrl: URL = new URL(
    "https://gist.githubusercontent.com/awharn/629aa52801a9a5f8b7f725b33572acf8/raw/79dafd7c98e53f10eb668a29ddeb08ae5412d609/zowe.config.json"
);
const schemaUrl: URL = new URL(
    "https://gist.githubusercontent.com/awharn/629aa52801a9a5f8b7f725b33572acf8/raw/79dafd7c98e53f10eb668a29ddeb08ae5412d609/zowe.schema.json"
);
const badConfigUrl: URL = new URL(
    "https://gist.githubusercontent.com/awharn/629aa52801a9a5f8b7f725b33572acf8/raw/79dafd7c98e53f10eb668a29ddeb08ae5412d609/zowe.config.json.bad"
);
const badSchemaUrl: URL = new URL(
    "https://gist.githubusercontent.com/awharn/629aa52801a9a5f8b7f725b33572acf8/raw/34cf180414061107ddb8b7f5a4e693b8fd7c2853/zowe.schema.json.bad"
);
const badUrl: URL = new URL(
    "https://gist.githubusercontent.com/awharn/629aa52801a9a5f8b7f725b33572acf8/raw/34cf180414061107ddb8b7f5a4e693b8fd7c2854/zowe.config.json.bad"
);

describe("Configuration import command handler", () => {

    describe("fetch config", () => {
        it("should be able to fetch the configuration file from a web address", async () => {
            const fetchConfig = (ImportHandler.prototype as any).fetchConfig;
            const config = await fetchConfig(configUrl);
            expect(config).toMatchSnapshot();
        });

        it("should handle errors encountered for an invalid JSON file", async () => {
            const fetchConfig = (ImportHandler.prototype as any).fetchConfig;
            let config: any;
            let error: any;
            try {
                config = await fetchConfig(badConfigUrl);
            } catch (err) {
                error = err;
            }

            expect(config).toBeUndefined();
            expect(error).toBeDefined();
            expect(error.message).toContain("unable to parse config");
        });

        it("should handle errors encountered for an invalid URL", async () => {
            const fetchConfig = (ImportHandler.prototype as any).fetchConfig;
            let config: any;
            let error: any;
            try {
                config = await fetchConfig(badUrl);
            } catch (err) {
                error = err;
            }

            expect(config).toBeUndefined();
            expect(error).toBeDefined();
            expect(error.message).toContain("Rest API failure with HTTP(S)");
        });
    });

    describe("download schema", () => {
        afterEach(() => {
            try {
                fs.unlinkSync(localDownloadedSchemaPath);
            } finally {
                // do nothing
            }
        });
        it("should be able to copy the schema file from a local file", async () => {
            const downloadSchema = (ImportHandler.prototype as any).downloadSchema;
            await downloadSchema(localSchemaUrl, localDownloadedSchemaPath);
            await flushPromises();
            const originalSchema = fs.readFileSync(localSchema).toString();
            const copiedSchema = fs.readFileSync(localDownloadedSchemaPath).toString();
            expect(copiedSchema).toEqual(originalSchema);
        });

        it("should be able to download the schema file from a web address", async () => {
            const downloadSchema = (ImportHandler.prototype as any).downloadSchema;
            await downloadSchema(schemaUrl, localDownloadedSchemaPath);
            await flushPromises();
            const originalSchema = fs.readFileSync(localSchema).toString();
            const copiedSchema = fs.readFileSync(localDownloadedSchemaPath).toString();
            expect(copiedSchema).toMatchSnapshot();
            expect(copiedSchema).toEqual(originalSchema);
        });

        it("should handle errors encountered for an invalid URL", async () => {
            const downloadSchema = (ImportHandler.prototype as any).downloadSchema;
            await downloadSchema(badSchemaUrl, localDownloadedSchemaPath);
            await flushPromises();
            const originalSchema = fs.readFileSync(localSchema).toString();
            const copiedSchema = fs.readFileSync(localDownloadedSchemaPath).toString();
            expect(copiedSchema).not.toEqual(originalSchema);
        });

        it.skip("should handle errors encountered for an invalid file path", async () => {
            const fetchConfig = (ImportHandler.prototype as any).fetchConfig;
            let config: any;
            let error: any;
            try {
                config = await fetchConfig(badUrl);
            } catch (err) {
                error = err;
            }

            expect(config).toBeUndefined();
            expect(error).toBeDefined();
            expect(error.message).toContain("Rest API failure with HTTP(S)");
        });

        it.skip("should handle errors encountered for an invalid output path", async () => {
            const fetchConfig = (ImportHandler.prototype as any).fetchConfig;
            let config: any;
            let error: any;
            try {
                config = await fetchConfig(badUrl);
            } catch (err) {
                error = err;
            }

            expect(config).toBeUndefined();
            expect(error).toBeDefined();
            expect(error.message).toContain("Rest API failure with HTTP(S)");
        });
    });
});