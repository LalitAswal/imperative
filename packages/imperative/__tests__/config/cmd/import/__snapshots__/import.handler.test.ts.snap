// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration import command handler download schema should be able to download the schema file from a web address 1`] = `
"{
    \\"$schema\\": \\"https://json-schema.org/draft/2020-12/schema\\",
    \\"$version\\": 3,
    \\"type\\": \\"object\\",
    \\"description\\": \\"config\\",
    \\"properties\\": {
        \\"profiles\\": {
            \\"type\\": \\"object\\",
            \\"description\\": \\"named profiles config\\",
            \\"patternProperties\\": {
                \\"^\\\\\\\\S*$\\": {
                    \\"type\\": \\"object\\",
                    \\"description\\": \\"a profile\\",
                    \\"properties\\": {
                        \\"type\\": {
                            \\"description\\": \\"the profile type\\",
                            \\"type\\": \\"string\\"
                        },
                        \\"properties\\": {
                            \\"description\\": \\"the profile properties\\",
                            \\"type\\": \\"object\\"
                        },
                        \\"profiles\\": {
                            \\"description\\": \\"additional sub-profiles\\",
                            \\"type\\": \\"object\\",
                            \\"$ref\\": \\"#/properties/profiles\\"
                        },
                        \\"secure\\": {
                            \\"description\\": \\"secure property names\\",
                            \\"type\\": \\"array\\",
                            \\"prefixItems\\": {
                                \\"type\\": \\"string\\"
                            },
                            \\"uniqueItems\\": true
                     "
`;

exports[`Configuration import command handler fetch config should be able to fetch the configuration file from a web address 1`] = `
Object {
  "$schema": "./zowe.schema.json",
  "defaults": Object {
    "base": "base",
    "ssh": "ssh",
    "tso": "tso",
    "zosmf": "zosmf",
  },
  "profiles": Object {
    "base": Object {
      "properties": Object {
        "host": "zowe.org",
        "password": "zowe",
        "rejectUnauthorized": true,
        "user": "zowe",
      },
      "secure": CommentArray [],
      "type": "base",
    },
    "ssh": Object {
      "properties": Object {
        "port": 22,
      },
      "secure": CommentArray [],
      "type": "ssh",
    },
    "tso": Object {
      "properties": Object {
        "account": "IZUACCT",
        "codePage": "1047",
        "logonProcedure": "IZUFPROC",
      },
      "secure": CommentArray [],
      "type": "tso",
    },
    "zosmf": Object {
      "properties": Object {
        "port": 443,
      },
      "secure": CommentArray [],
      "type": "zosmf",
    },
  },
}
`;
